### ISSUE TITLE
clean old data to be GDPR compliant

### ISSUE THREAD

**Issue by npracht** [mautic/mautic#5878]
| Q  | A
| ---| ---
| Bug report? | 
| Feature request? | 
| Enhancement? | x

## Description:
With the GDPR coming in May 2018, it is mandatory for all companies to completely remove data from inactive contacts after 3 years (maximum) of inactivity.

@alanhartless before starting working on that, i'd like your opinion because i think we can get inspired from https://github.com/mautic/mautic/pull/2024.

## Suggestion
Add a parameter add the end of the `php app/console mautic:maintenance:cleanup --years-old=3` command like `--gdpr` that uses last active date instead of creation date and that removes all contact data (not only anonymous).

Your opinion ? :)
Thanks !

**Comment by alanhartless** [comment]
The command currently deletes based on last active date but it does limit to visitors/anonymous. I don't see any reason why not to add a `--gdpr` flag.

**Comment by npracht** [comment]
Label: WIP

**Comment by dbhurley** [comment]
+1

**Comment by heathdutton** [comment]
I assume a country filter will be used so that you are only filtering EU contacts?
I could see the need to specify which countries to include/exclude in the cleanup. Some may opt to include additional countries for other corporate policies.

**Comment by npracht** [comment]
@heathdutton if the mautic user is in EU he should do it for all the contacts. Not only contacts in EU.

**Comment by escopecz** [comment]
I was answering [interesting question](https://mautic.slack.com/archives/C02HU8BUT/p1523870205000476) for this topic today. The thing is that you can configure it with Mautic current features even today. No need to develop anything new.

1. Create a segment with these filters:
![screen shot 2018-04-16 at 16 47 22](https://user-images.githubusercontent.com/1235442/38816228-e8d218f2-4195-11e8-9ac6-cf95620077fb.png)

2. Create a campaign to delete all contacts in this segment:
![screen shot 2018-04-16 at 16 48 46](https://user-images.githubusercontent.com/1235442/38816297-10dccdd8-4196-11e8-9dc2-471b0f2bbe97.png)

If I understand GDPR correctly, you can have data older than 3 years if they are anonymised. So this will work well, right? You can have page hits that you still see in your stats, but there won't be any existing contact who created them.

A problem may be in the form submissions or if a page/email hit contains a contact information. In this case a new code is needed to remove that.

Another idea for a new feature would be an anonymizer. Such feature would not delete all data, but hash the sensitive data instead. This way Mautic could still track the same contact, but a Mautic user could see only non-sensitive data about that contact.

**Comment by npracht** [comment]
@escopecz the campaign works. Marketingly it would be better if it is just a question of setting/setup (CRONs config are inside of it for me).

You **do need** to delete them if inactive for more than 3 years. Even if it is stupid it is said like that ---> in France for instance you are obligated to keep your invoices for 5 years. If your customer is inactive for more than 3 years.... 🔫 

On my opinion the option `--gdpr` should simply remove all the data related to a contact ---> form submission, tracked pages, etc.
I'd say that you don't make statistics on datas aged of 3 years. You mainly check previous year and current year data.
______

Conclusion: What do you think? Should i go to that option or you won't agree to merge that and you prefer another option?
Thanks a lot for your time and help!

**Comment by escopecz** [comment]
Alan and DB already gave you 👍. So I'd consider it approved.

**Comment by YosuCadilla** [comment]
> Another idea for a new feature would be an anonymizer.

**Comment by YosuCadilla** [comment]
Are you sure you need to delete ALL data and not only the data which could be directly related to a person or other personal data?

**Comment by heathdutton** [comment]
Just a caution: The property last_active is allowed to be NULL for the sake of imported contacts from other systems/integrations/etc. So for contacts that have never been active on your site (as in a registered hit), you'll want to fall back on date identified (or last modified).

**Comment by npracht** [comment]
@heathdutton & @YosuCadilla PR is here #6242 !

**Comment by npracht** [comment]
Merged into 2.15.0
